plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/TypeEmbedding.kt

```diff
@@ -251,5 +261,49 @@ data class ClassTypeEmbedding(val className: ScopedKotlinName) : TypeEmbedding {
     override fun accessInvariants(v: Exp): List<Exp> =
         flatMapUniqueFields { _, field -> field.accessInvariantsForParameter(v) }
 
+    // Note: this function will replace accessInvariants when nested unfold will be implemented
+    override fun predicateAccessInvariants(v: Exp): List<Exp> = listOf(Exp.PredicateAccess(name, listOf(v)))
+
     override fun provenInvariants(v: Exp) = listOf(subtypeInvariant(v))
+
+    private val thisEmbedding = VariableEmbedding(ThisReceiverName, this)
+    val classPredicate: Predicate
+        get() {
+            val accessReadableFields = fields.values
+                .filter { it.accessPolicy == AccessPolicy.ALWAYS_READABLE }
+                .map { FieldAccess(thisEmbedding, it).getAccessPredicate(PermExp.WildcardPerm()) }
+
+            val accessSpecialFields = fields.values
+                .filter { it.accessPolicy == AccessPolicy.ALWAYS_WRITEABLE }
+                .map { FieldAccess(thisEmbedding, it).getAccessPredicate(PermExp.FullPerm()) }
+
+            val classFieldPredicates = fields.values
+                .filter { it.accessPolicy == AccessPolicy.ALWAYS_READABLE }
+                .map { it.type.predicateAccessInvariants(Exp.FieldAccess(Exp.LocalVar(ThisReceiverName, Type.Ref), it.toViper())) }
+                .flatten()
+
+            val superTypesPredicates = superTypes
+                .filterIsInstance<ClassTypeEmbedding>()
+                .map { Exp.PredicateAccess(it.name, listOf(thisEmbedding.toViper())) }
+
+            val conditions = accessReadableFields + accessSpecialFields + specialConditions + classFieldPredicates + superTypesPredicates
+            val body = if (conditions.isNotEmpty()) {
+                conditions
+                    .drop(1)
+                    .fold(conditions.first()) { acc, expEmbedding -> Exp.And(acc, expEmbedding) }
```

Comment:
See my recent commit for this.  (I'll modify it to also allow the empty case, since it'll be useful for you.)